libs/@local/harpc/client/typescript/package.json

```diff
@@ -0,0 +1,26 @@
+{
+  "name": "@local/harpc-client",
+  "version": "0.0.0-private",
+  "private": true,
+  "description": "A collection of helper types",
+  "license": "AGPL-3",
+  "type": "module",
+  "exports": {
+    ".": {
+      "import": "./dist/index.js",
+      "types": "./dist/index.d.ts"
+    }
+  },
+  "scripts": {
+    "build": "tsup src/index.ts",
```

Comment:
Is it a drop-in replacement? If so doesn't it make sense to use it elsewhere as well? cc @CiaranMn

Reply:
It is yes, you can customize the build process as well via different configuration files (https://tsup.egoist.dev/#using-custom-configuration) - but tsup is build to work out of the box with sane defaults.

Reply:
We mostly use `tsc` for type checking. `tsup` is partly faster because it does no type checking. You need to add the `--dts` flag to get it to check types. There are a [few open issues](https://github.com/egoist/tsup/issues?q=is%3Aissue+is%3Aopen+--dts) related to this mode, some of which don't make me comfortable that it would be performing the same level of type checking / respect our config, even if it was faster with `--dts` enabled. So it's worth further investigation at some point but I don't think an easy decision to swap.

Reply:
From the docs:

> **What about type checking?**
> esbuild is fast because it doesn't perform any type checking, you already get type checking from your IDE like VS Code or WebStorm.
> Additionally, if you want type checking at build time, you can enable --dts, which will run a real TypeScript compiler to generate declaration file so you get type checking as well.

'you already get type checking from your IDE' doesn't inspire me with confidence that it's a good replacement for what we mostly use `tsc` for (making sure there are no type errors in CI).

Reply:
Well, my usual workflow is a bit different - instead of typechecking every single build - you only typecheck on a production build. That way you can keep the iteration time high until you hit prod.

`tsup` is faster because yes - it simply strips the types (and is written in Go) - the stipulation here is that your editor is going to find those issues for you when you work on it and then the prod build is your last "sanity" check. (or in our case that could be CI)

Reply:
So you wouldn't completely replace tsc, you would simply do a full typecheck in CI instead (/prod build). This is also how tools like `vite` get such a significant speedup.

Reply:
Right, but we're not building anything locally, we're running the TS directly (via `tsx`), `tsc` only runs in CI (or if you manually run linting). So there's not really any development/iteration uses of `tsc` to replace. I do want to switch to building JS locally at some point because the current setup is too complicated and relies on maintaining `paths` aliases in the TS config alongside the `exports` in each package. At which point using `tsup` looks worthwhile for that purpose.

Reply:
I see, I didn't know that. I thought that whenever I update e.g. `@local/harpc-client` I'd need to rebuild it in our setup to make use of it in `hash-api`. My mistake. So instead I should amend the tsconfig to provide the additional path I guess?

Reply:
see: [3e43235](https://github.com/hashintel/hash/pull/5620/commits/3e43235913600c935956b96fdb946bbf32236e70)

Reply:
Thank you – can you also add the alias to `paths` in [this file](https://github.com/hashintel/hash/blob/main/libs/%40local/tsconfig/legacy-base-tsconfig-to-refactor.json)? 

I think using `tsup` for building is a good idea, we're just not building properly at the moment. We could still use `tsup` as part of the `build` command since types are checked in `tsc` anyway (so probably I was wrong to say there's no uses of `tsc` to replace – we can do so if `tsup` otherwise builds things without any issues to do with config).

Reply:
I added it: [3f487f6](https://github.com/hashintel/hash/pull/5620/commits/3f487f6f6d7057fd381f6b63fcf848d47a8fc4af)

I was kinda on autopilot when introducing `tsup` (because I make use of it in all of my projects), so was definitely wrong to introduce here first.

Maybe it'd make sense to revisit this at a later date, as this feels like the wrong place to do it in. My bad, sorry!

This way we'd have a `lint:tsc` pass, then a build step via `tsup` and won't need to duplicate the work.

Reply:
Definitely worth revisiting, and I appreciate the suggestion, as I didn't know about `tsup`.

I think linting will complain about `index.ts` btw, so would rename to `main.ts`

The original rationale for this lint was so that it was clear if you do

`import "./src"` 

it's clear you are importing a file called `./src.js` and not `./src/index.js`

Because we enforce import extensions now, this is pretty much redundant, although it's not worth renaming the many `main` files back to `index` (unless there's a benefit to it I haven't thought of, which is quite possible)

Reply:
I have disabled the lint for that specific file (https://github.com/hashintel/hash/blob/3f487f6f6d7057fd381f6b63fcf848d47a8fc4af/libs/%40local/harpc/client/typescript/src/index.ts#L1), while I don't mind renaming to `main.ts`, especially for the entrypoint I like to still use `index.ts`, the reason is intent. The `index.ts` is the index into the package itself, while (at least in my mind) `main.ts` conveys more of a this is the primary content of the package. My goal here is to have `index.ts` be only a place of re-exports so it's more of an index into the package than the main part of the package.

I can also revert that change if needed.

Reply:
No, that's fine, the lint is redundant now anyway